### ISSUE TITLE
[JENKINS-40805] Workspace not available when a slave is created by the Kubernetes plugin

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#1843]
<p>I'm configuring an environment with a jenkins master that uses kubernetes to instantiate slaves.<br/>
Everyting works fine except that the workspace is not available when the slave is killed by the kubernetes (slave offline).<br/>
I have tried several solutions (using an NFS, copying back the workspace on the master with the copy-to-slave plugin, ... aso) .... but it doesn't work ... Jenkins don't want to display the workspace if the node is offline (even if the workspace is at the right location, but on the master).<br/>
Do you have a solution ?</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivier_lambert">olivier_lambert</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805" target="_blank">Workspace not available when a slave is created by the Kubernetes plugin</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: kubernetes-plugin
<li><b>resolution</b>: Not A Defect
<li><b>resolved</b>: 2021-01-08T09:08:17+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 16
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I'm configuring an environment with a jenkins master that uses kubernetes to instantiate slaves.
Everyting works fine except that the workspace is not available when the slave is killed by the kubernetes (slave offline).
I have tried several solutions (using an NFS, copying back the workspace on the master with the copy-to-slave plugin, ... aso) .... but it doesn't work ... Jenkins don't want to display the workspace if the node is offline (even if the workspace is at the right location, but on the master).
Do you have a solution ?</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [PR-142](https://github.com/jenkinsci/kubernetes-plugin/pull/142)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rpallikonda:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=285254&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285254">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a>
  Could you please help with this issue, or how we can fix it?</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivier_lambert" class="user-hover" rel="olivier_lambert">olivier_lambert</a> 
  Could you please let me know if you were able to solve the issue?</p>

<p>We do have a workaround which is not that nice.. Once the build is done, we have to change the 'where the project is run ' to master. Then the workspace is shown.</p>

</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a><br/>
  Could you please help with this issue, or how we can fix it?</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivier_lambert" class="user-hover" rel="olivier_lambert">olivier_lambert</a> <br/>
  Could you please let me know if you were able to solve the issue?</p>

<p>We do have a workaround which is not that nice.. Once the build is done, we have to change the 'where the project is run ' to master. Then the workspace is shown.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>olivier_lambert:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=285265&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285265">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpallikonda" class="user-hover" rel="rpallikonda">rpallikonda</a>
We also find this workaround. It works fine .... but it's just a workaround ...
We think a better solution should be implemented directlry in the core of Jenkins.
1. Check if the "where the project is run" is online
2. IF YES &gt;&gt; display this workspace
3. IF NO &gt;&gt; check if it's available on the master
4. IF YES, display it
5. IF NO, display error message</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpallikonda" class="user-hover" rel="rpallikonda">rpallikonda</a><br/>
We also find this workaround. It works fine .... but it's just a workaround ...<br/>
We think a better solution should be implemented directlry in the core of Jenkins.<br/>
1. Check if the "where the project is run" is online<br/>
2. IF YES >> display this workspace<br/>
3. IF NO >> check if it's available on the master<br/>
4. IF YES, display it<br/>
5. IF NO, display error message</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>anhu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=299652&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299652">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I want to support the suggestion of <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivier_lambert" class="user-hover" rel="olivier_lambert">olivier_lambert</a> to display the workspace from the master if the slave is gone.</p>

<p>We are using the jenkins swarm plugin and create slaves dynamically. An NFS server ensures that all workspace data is available on the master. However, every time a slave gets removed the UI cannot display the workspaces of its jobs any more. We even tried to create our own WorkspaceBrowser extension, but this did not work reliably and we could not pin down the exact issue.</p>

<p>A simple check if the workspace is available on the same location on the master and use this path could fix this issue.</p>
</pre>
</details>

</ul>
</details>

<p>I want to support the suggestion of <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivier_lambert" class="user-hover" rel="olivier_lambert">olivier_lambert</a> to display the workspace from the master if the slave is gone.</p>

<p>We are using the jenkins swarm plugin and create slaves dynamically. An NFS server ensures that all workspace data is available on the master. However, every time a slave gets removed the UI cannot display the workspaces of its jobs any more. We even tried to create our own WorkspaceBrowser extension, but this did not work reliably and we could not pin down the exact issue.</p>

<p>A simple check if the workspace is available on the same location on the master and use this path could fix this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rpallikonda:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=299918&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299918">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=anhu" class="user-hover" rel="anhu">anhu</a> Hi Andreas, please check the PR <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/142" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/142</a> which has the WorkspaceBrowser fix that solves the issue. </p>

<p>In our case, we do not have a NFS which is available to both master &amp; slaves, ie. the PVC is only attached to master. We have the 'copy to slave' plugin post-build action to have the workspace copied to master, and the above PR helps with making it browsable from master even if the slave is gone.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=anhu" class="user-hover" rel="anhu">anhu</a> Hi Andreas, please check the PR <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/142" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/142</a> which has the WorkspaceBrowser fix that solves the issue. </p>

<p>In our case, we do not have a NFS which is available to both master & slaves, ie. the PVC is only attached to master. We have the 'copy to slave' plugin post-build action to have the workspace copied to master, and the above PR helps with making it browsable from master even if the slave is gone.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>anhu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=299929&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299929">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thank you for your comment, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpallikonda" class="user-hover" rel="rpallikonda">rpallikonda</a> . I tried to adapt my WorkspaceBrowser extension with the info i got from the PR. I am quite confident, that my problem does not lie with my WorkspaceBrowser anymore.</p>

<p>However, on some jobs my WorkspaceBrowser extension is simply not called. But this seems to be a total different issue, and has nothing to do with cloud/swarm plugins. If the extension is called, its working as expected. Deleting the builds/ folder and reloading the config seems to be a workaround.</p>
</pre>
</details>

</ul>
</details>

<p>Thank you for your comment, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpallikonda" class="user-hover" rel="rpallikonda">rpallikonda</a> . I tried to adapt my WorkspaceBrowser extension with the info i got from the PR. I am quite confident, that my problem does not lie with my WorkspaceBrowser anymore.</p>

<p>However, on some jobs my WorkspaceBrowser extension is simply not called. But this seems to be a total different issue, and has nothing to do with cloud/swarm plugins. If the extension is called, its working as expected. Deleting the builds/ folder and reloading the config seems to be a workaround.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/lionelve.png" title="lionelve's avatar" /> lionelve:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=311016&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311016">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpallikonda" class="user-hover" rel="rpallikonda">rpallikonda</a> Can I ask what is this "where the project is run" property you use for your workaround?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpallikonda" class="user-hover" rel="rpallikonda">rpallikonda</a> Can I ask what is this "where the project is run" property you use for your workaround?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rpallikonda:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=312646&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-312646">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=lionelve" class="user-hover" rel="lionelve">lionelve</a> Apologize the delay in the reply.  In our case, we use the 'copy to slave' plugin to have the workspace copied back to master as a "post build step" in the jenkins job.</p>

<p> </p>

<p>When users access the workspace after the slave is offline, then the above PR would redirect the Jenkins to the workspace path on the "master".</p>

<p>The "where the project is run" is the label of the "pod template".</p>

<p>Jenkins uses the k8s plugin's workspace browser when the job's "where the project is run" is associated with the k8s plugin.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=lionelve" class="user-hover" rel="lionelve">lionelve</a> Apologize the delay in the reply.  In our case, we use the 'copy to slave' plugin to have the workspace copied back to master as a "post build step" in the jenkins job.</p>

<p> </p>

<p>When users access the workspace after the slave is offline, then the above PR would redirect the Jenkins to the workspace path on the "master".</p>

<p>The "where the project is run" is the label of the "pod template".</p>

<p>Jenkins uses the k8s plugin's workspace browser when the job's "where the project is run" is associated with the k8s plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jbounds:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=345901&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345901">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivier_lambert" class="user-hover" rel="olivier_lambert">olivier_lambert</a>, looks like this PR was never approved. How did you get around this issue? <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a> We have the same exact issue. Looks like the PR was never approved. Can we submit a new PR for this?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivier_lambert" class="user-hover" rel="olivier_lambert">olivier_lambert</a>, looks like this PR was never approved. How did you get around this issue? <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a> We have the same exact issue. Looks like the PR was never approved. Can we submit a new PR for this?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> csanchez:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=345913&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345913">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You can continue with the existing PR it was never completed</p>
</pre>
</details>

</ul>
</details>

<p>You can continue with the existing PR it was never completed</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jbounds:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=345914&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345914">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Awesome. Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a>.</p>
</pre>
</details>

</ul>
</details>

<p>Awesome. Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ffarah:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=350710&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-350710">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a>, using copy-to-slave plugin is a terrible idea. Because that plugin isn't maintained and is even taken off of the official Jenkins plugin list due to unfixed security problems. Being able to view the workspace after the slave node is destroyed is quite important and would be very useful to have that feature built in.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez" class="user-hover" rel="csanchez">csanchez</a>, using copy-to-slave plugin is a terrible idea. Because that plugin isn't maintained and is even taken off of the official Jenkins plugin list due to unfixed security problems. Being able to view the workspace after the slave node is destroyed is quite important and would be very useful to have that feature built in.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tdihp:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=356687&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-356687">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It seems post-build actions in freestyle build, that uses workspace, such as S3 Publisher, will also fail trying to connect to the pod by JNLP.</p>
</pre>
</details>

</ul>
</details>

<p>It seems post-build actions in freestyle build, that uses workspace, such as S3 Publisher, will also fail trying to connect to the pod by JNLP.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/yrsurya.svg" title="yrsurya's avatar" /> yrsurya:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=359671&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-359671">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>May be we use of <a href="https://github.com/jenkinsci/external-workspace-manager-plugin/blob/master/README.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/external-workspace-manager-plugin/blob/master/README.md</a> this plugin to get view the workspace form GUI if some configurations changes happen in both kubernetes plugin and as well Jenkins</p>
</pre>
</details>

</ul>
</details>

<p>May be we use of <a href="https://github.com/jenkinsci/external-workspace-manager-plugin/blob/master/README.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/external-workspace-manager-plugin/blob/master/README.md</a> this plugin to get view the workspace form GUI if some configurations changes happen in both kubernetes plugin and as well Jenkins</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/yrsurya.svg" title="yrsurya's avatar" /> yrsurya:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=369161&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-369161">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can I know if any one able to solve this issue??</p>
</pre>
</details>

</ul>
</details>

<p>Can I know if any one able to solve this issue??</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nnalagesigari:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=384346&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-384346">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am also facing the issue where we would like to see the job's workspace after the pod terminated . Our agents are attached to a PVC ( persistence volume ) and can still see the workspace even after the pod terminates but cannot access through the job bcz no pod exist. </p>

<p>please let me know if you found ways to address this issue </p>
</pre>
</details>

</ul>
</details>

<p>I am also facing the issue where we would like to see the job's workspace after the pod terminated . Our agents are attached to a PVC ( persistence volume ) and can still see the workspace even after the pod terminates but cannot access through the job bcz no pod exist. </p>

<p>please let me know if you found ways to address this issue </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vlatombe.png" title="vlatombe's avatar" /> vlatombe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=403115&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-403115">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is a generic problem with every cloud implementation on Jenkins afaiu. The solution won't come from a particular implementation (kubernetes plugin in this case). Retrieving the workspace back to the master has a cost, which can be pretty high.</p>

<p>The initial PR linked to this ticket didn't really do anything, and it was only trying to cover freestyle projects, not pipelines.</p>

<p>If you want to keep something for later, you need to archive it.</p>
</pre>
</details>

</ul>
</details>

<p>This is a generic problem with every cloud implementation on Jenkins afaiu. The solution won't come from a particular implementation (kubernetes plugin in this case). Retrieving the workspace back to the master has a cost, which can be pretty high.</p>

<p>The initial PR linked to this ticket didn't really do anything, and it was only trying to cover freestyle projects, not pipelines.</p>

<p>If you want to keep something for later, you need to archive it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tavin:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-40805?focusedId=403124&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-403124">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Nevertheless, if we do in fact mount the same filesystem for workspaces on the master as on the agents (using NFS for example), then jenkins should be smart enough to let you browse the workspace. Agreed it's not a kubernetes-plugin issue, but there is still a defect here.</p>
</pre>
</details>

</ul>
</details>

<p>Nevertheless, if we do in fact mount the same filesystem for workspaces on the master as on the agents (using NFS for example), then jenkins should be smart enough to let you browse the workspace. Agreed it's not a kubernetes-plugin issue, but there is still a defect here.</p>
